Edge Coloring for bipartite graphs

padath314 opened the pull request:
The function `bipartite_edge_coloring()` determines the minimum edge coloring for a bipartite graph, employing two strategies: iterated matching and Kempe chain. The default strategy is Kempe chain. Five graphs were used to test both strategies.

The function returns a dictionary where the edges serve as keys, and the corresponding values represent the assigned colors.

padath314 commented:
@dschult The tests are running fine in our system. Could you please guide us on why it is failing in dispatch? 
I've also opened a conversation in the mailing list explaining our work.

padath314 commented:
@dschult Could you please check this out. I've also opened a [conversation](https://groups.google.com/g/networkx-discuss/c/kUtTJTL4cwA/m/n08moYfkAAAJ) in the mailing list explaining our work.

dschult commented on networkx/algorithms/bipartite/tests/test_edgecoloring.py:
```diff
@@ -0,0 +1,227 @@
+        used_colors = set(coloring.values())
+        assert len(used_colors) == max_degree
+
+    # def test_odd_cycle(self):
```
just a reminder that this looks anticipated... is it still?

Sreelaxme replied:
Fixed it

dschult reviewed (commented):
Thanks for the bipartite edge_coloring algorithms. 
I don't know much about this area, so it'll be fun to get to know it a little bit. :)

I have a couple of style-oriented comments apppear below.  
Here's some Big Picture stuff:
- you will need to connect this to the import structure of the library by adding a line to the `__init__.py` file in the bipartite directory.
- You will need to create a documentation section in `doc/reference/algorithms/bipartite.rst` for this work so the doc_strings get included in the reference section of the documentation. You can check what the documentation looks like by clicking on the "Details" link in the CI check called "documentation" (or build it locally if you prefer by going to `doc/` and then `make html` after installing the requirements for building docs).
- Why not make two functions (one for each strategy) instead of the third function that calls the other two?  Do you expect other strategies later? It also looks like the strategies return "the minimum matching", while the main function returns "a valid matching". Why that distinction? What is a "minimum" matching?  etc?
- For helper functions, if the helper is only used once and isn't too crazy it is usually better to in-line the function for readability (and sometimes for performance). Otherwise the person has to go lookup the helper separately.
- You should probably only make copies of the graph when needed (and maybe you do need to here, but I don't see where). Also, you remove edges from G -- and maybe this is where I'm missing the need for a copy -- but it doesn't seem like you need to do that. Could you use a subgraph instead, or maybe a list of edges yet to be colored?

padath314 commented:
@dschult Thankyou for your feedback. We're working to rectify everything ASAP.

padath314 commented on networkx/algorithms/bipartite/edge_coloring.py:
```diff
@@ -0,0 +1,351 @@
+
+    Raises
+    ------
+    NetworkXError
+        Raised if the input graph is not bipartite.
+
+    AmbiguousSolution
+      Raised if the input bipartite graph is disconnected and no container
+      with all nodes in one bipartite set is provided. When determining
```
I have changed this. thankyou for pointing it out.

padath314 replied:
I have changed this. thankyou for pointing it out.

padath314 commented:
> Thanks for the bipartite edge_coloring algorithms. I don't know much about this area, so it'll be fun to get to know it a little bit. :)
> 
> I have a couple of style-oriented comments apppear below. Here's some Big Picture stuff:
> 
> * you will need to connect this to the import structure of the library by adding a line to the `__init__.py` file in the bipartite directory.
> * You will need to create a documentation section in `doc/reference/algorithms/bipartite.rst` for this work so the doc_strings get included in the reference section of the documentation. You can check what the documentation looks like by clicking on the "Details" link in the CI check called "documentation" (or build it locally if you prefer by going to `doc/` and then `make html` after installing the requirements for building docs).
> * Why not make two functions (one for each strategy) instead of the third function that calls the other two?  Do you expect other strategies later? It also looks like the strategies return "the minimum matching", while the main function returns "a valid matching". Why that distinction? What is a "minimum" matching?  etc?
> * For helper functions, if the helper is only used once and isn't too crazy it is usually better to in-line the function for readability (and sometimes for performance). Otherwise the person has to go lookup the helper separately.
> * You should probably only make copies of the graph when needed (and maybe you do need to here, but I don't see where). Also, you remove edges from G -- and maybe this is where I'm missing the need for a copy -- but it doesn't seem like you need to do that. Could you use a subgraph instead, or maybe a list of edges yet to be colored?

In commit  - [1b14a3f](https://github.com/networkx/networkx/pull/7138/commits/1b14a3f008197ccf55141b55a388dde987ce8fcd)
1. Connected to the import structure
2. Created a documentation section
3.We chose to implement a wrapper function which will pick a strategy by default since
    - There are few more strategies (slightly more complicated ones) available in literature and with slightly better running times.
    - We wanted the user to be able to do an edge-coloring using a standard function name like "bipartite_edge_coloring" without worrying which strategy is being used. 
  But if you think the wrapper function is not necessary, we can avoid that and move the exception handlings to each of the two specific functions.
  About the main functioning returning a 'valid matching' instead a 'minimum' matching, it is a typo. Fixed it now.

padath314 commented on networkx/algorithms/bipartite/tests/test_edgecoloring.py:
```diff
@@ -0,0 +1,227 @@
+        u, v = edge  # Assuming the edges are represented as pairs (u, v)
+
+        if u not in vertex_colors:
+            vertex_colors[u] = set()
```
There is no need to add  both u and v here because the coloring returned will have both keys (u, v) and (v, u) in it. I have modified the function to use defaultdict.

Sreelaxme commented on networkx/algorithms/bipartite/edge_coloring.py:
```diff
@@ -0,0 +1,351 @@
+
+import networkx as nx
+from networkx.generators.classic import null_graph
+from networkx.utils import not_implemented_for
```
Resolved

Sreelaxme commented on networkx/algorithms/bipartite/edge_coloring.py:
```diff
@@ -0,0 +1,351 @@
+
+    References:
+    -----------
+    Harold N.Gabow and Oded Kariv , “Algorithms for edge coloring bipartite graphs”
```
I have added the full reference with link

rossbar reviewed (commented):
Just a couple quick comments from a *very* rapid readthrough - I'll have to take a closer look in the future.

Re: the proposed API, I think both approaches (i.e. wrapper function with methods, or individual functions for each method) have their merits. If we go with the wrapper function approach, I'd vote to then make the functions that implement the individual methods private (i.e. add an underscore in front of `kempe` and `iterated` and don't add them to `__all__`). That way, there's only one interface exposed to the user and no ambiguity about how it should be used.

I'd also vote to remove the `bipartite` from the function name, as it would be redundant with the recommended calling pattern: `nx.bipartitie.edge_coloring` is more clear/concise IMO.

padath314 commented:
Thank you, @rossbar and @dschult  for your feedbacks. We have addressed the concerns raised and made necessary adjustments. 

However, we have not yet inlined the helper functions _combined_matchings and _matching_saturating_max_degree. The hesitation stems from the fact that the former function is lengthy, and inlining it might negatively impact code readability.  Your insights on this matter would be greatly appreciated.

Sreelaxme commented:
Hello @rossbar  and @dschult , just a friendly reminder about the pending decision on inlining the helper functions _combined_matchings and _matching_saturating_max_degree. Your insights on this matter would be immensely helpful in finalizing this pull request. Looking forward to your thoughts whenever you get a chance!

padath314 commented:
@rossbar @dschult 
We kindly request your immediate assistance with our Undergraduate project. As we are nearing graduation, your prompt action would be greatly appreciated. We are eager to enhance our code and progress to non-bipartite edge coloring. Your understanding of the urgency is highly valued.

dschult commented:
Hi Sorry for the delay. I have tried to think through which is better and on such short timescale I'm not sure I know which is better. What have you been taught about how to decide whether to split code into functions? Splitting makes each section shorter, but your future self may need to be able to look at the helper function while reading the primary bit of code. That can be a pain especially if the code gets moved farther apart in the file over time.  It probably doesn't matter much, but maybe you have been taught something about that choice.

Tell you what -- make the decision you think is best, let us know what that is and why and call the project done -- the term is over :)  I can't promise that we will review and merge it before your class finishes. But I certainly understand the desire to get the PR into good enough shape to call it "done".

padath314 commented:
@dschult @rossbar 
I've made the suggested change by in-lining the function to improve readability, considering that the function `combine_matchings` has a specific use case that likely won't be needed by others. However, the commit is failing checks due to changing `dispatch` to `dispatchable`. What should we do in this situation?

dschult commented:
I was most of the way to fixing the issues with dispatchable when I ran into configuration problems and haven’t been able to get back to it. I’m aiming for this weekend to get the dispatchable issue sorted out.

it is a matter of “rebase”ing or merging the PR with the main branch of NetworkX.  The two branches have diverged so we need to tell git how to handle merging them together.

dschult commented:
I've merged the PR with `main` and reapplied your PR to change the name of dispatch to dispatchable.
It looks like it is working -- hope for some more green checks! :)

My merge and push to your branch means that you will need to pull the current branch to your local repo before making any further changes.   * fingers crossed *

Sreelaxme commented:
All the checks are passing. Waiting for the approval.

dschult commented on networkx/algorithms/bipartite/edge_coloring.py:
```diff
@@ -0,0 +1,279 @@
+
+            u, v = edges[0]
+            if (len(edges) % 2 == 1) or (
+                (u in M1 and M1[u] == v and u in part_a)
+                or (u in M2 and M2[u] == v and u in part_b)
```
don't we know whether `u in part_a` based on the parity of len(edges)? Or said another way, how can u be in M1 and not be in part_a?  <similar for v in part_b>

dschult commented on networkx/algorithms/bipartite/edge_coloring.py:
```diff
@@ -0,0 +1,279 @@
+                final_matching[u1] = v1
+                final_matching[v1] = u1
+        elif v == u1:
+            for i in range(0, len(edges), 2):
+                u, v = edges[i]
```
Similar to 4 lines above -- couldn't you use:
```suggestion
            for u, v in edges[::2]:
```

dschult commented on networkx/algorithms/bipartite/tests/test_edgecoloring.py:
```diff
@@ -0,0 +1,225 @@
+    vertex_colors = defaultdict(set)
+
+    for edge, color in coloring.items():
+        u, _ = edge  # Assuming the edges are represented as pairs (u, v)
```
Is it sufficient to check only the head node of each edge to see the coloring is proper?
Also, this doesn't check that every edge in G is colored in coloring. Indeed, this function doesn't look at G at all.  In each test you use this function, but then you do some other standard computations to check if it is valid. Why not put it all into the tests? Why not put it all into this function? It doesn't really matter which, but it seems strange to do it halfway.

Could you go through all the new code and switch `vertex/vertices` to `node/nodes`? That is the NetworkX chosen term just for uniformity (and ease of spelling). And while we aren't 100% in this trait, it'd be good to introduce new code with that term.

padath314 replied:
It is sufficient as in the `coloring` dictionary, for every edge (u, v) there are two corresponding keys - (u, v) and (v, u). We followed this format from the `matching` code.

padath314 replied:
I have done the mentioned changes ie. checked all edges of the graph and also changed all the instances of vertex/vertices to node/nodes.

dschult commented on networkx/algorithms/bipartite/tests/test_edgecoloring.py:
```diff
@@ -0,0 +1,225 @@
+    return True
+
+
+class TestEdgeColoring_IteratedMatching:
+    """Tests for bipartite iterated matching edge coloring algorithm"""
+
```
Since all the tests for the two strategies are the same, you could simplify the code by using the `pytest.mark.parametrize` feature of pytest. See `networkx/algorithms/tests/test_mis.py` or the docs for pytest parametrize. Something like:

```python
@pytest.mark.parametrize("strategy", ["iterated-matching", "kempe_chain"])
class TestEdgeColoring_IteratedMatching:
    def test_complete_graph(self, strategy):
```

padath314 replied:
done.

dschult reviewed (commented):
This PR is getting there. Here are a bunch more comments/suggestions. I'm not saying you have to do them all -- but I'd like you to think about them and find the best way forward.  :)  Thanks!

padath314 commented on networkx/algorithms/bipartite/edge_coloring.py:
```diff
@@ -0,0 +1,279 @@
+import networkx as nx
+
+__all__ = ["edge_coloring"]
+
```
I have changed the module name to `edge_colorings` as you suggested.

padath314 replied:
Changed the file name to `edge_colorings`

padath314 commented on networkx/algorithms/bipartite/edge_coloring.py:
```diff
@@ -0,0 +1,279 @@
+@nx.utils.not_implemented_for("multigraph")
+@nx._dispatchable(name="bipartite_edge_coloring")
+def edge_coloring(G, top_nodes=None, strategy="kempe-chain"):
+    """Returns a minimum edge coloring of the bipartite graph `G`.
```
Yes, this is the coloring with minimum possible number of colors, hence naming the function as `minimum_edge_coloring` is better. But for convenience we have also added an alias `edge_coloring` for the same.

padath314 replied:
We changed the function name to `minimum_edge_coloring` as the name describes the function well. But for convenience we have also added an alias called `edge_coloring` for convenience.

eriknw replied:
> hence naming the function as `minimum_edge_coloring` is better. But for convenience we have also added an alias `edge_coloring` for the same.

I'm curious about best practice for adding aliases in NetworkX. I know `maximum_matching` is an alias for `hopcroft_karp_matching`, which is mentioned in their docstring. I don't know of other aliases, but I could probably find them programmatically.

I'm always a little bit surprised whenever I see aliases, and I expect an alias would be added when a function is heavily used (and names obvious) so as to justify the convenience. I'm not familiar with this algorithm, so my uninformed knee-jerk reaction would be to prefer to avoid aliasing this function and to choose one name such as `minimum_edge_coloring`.

padath314 commented on networkx/algorithms/bipartite/edge_coloring.py:
```diff
@@ -0,0 +1,279 @@
+    """Returns the minimum edge coloring of the bipartite graph `graph`.
+
+    This function uses the procedure augment to color the edges of the bipartite
+    graph such that no two adjacent edges have the same color.
```
`procedure-augment` is mentioned as it is in the reference we have mentioned. We have also written a short description.

padath314 commented on networkx/algorithms/bipartite/edge_coloring.py:
```diff
@@ -0,0 +1,279 @@
+    """Returns the minimum edge coloring of the bipartite graph `G`.
+
+    This function uses the procedure one-color to color the edges of the bipartite
+    graph such that no two adjacent edges have the same color.
```
Done.

padath314 commented on networkx/algorithms/bipartite/edge_coloring.py:
```diff
@@ -0,0 +1,279 @@
+    An edge coloring is an assignment of colors to the edges of a graph such
+    that no two adjacent edges share the same color. In the case of bipartite
+    graphs, the coloring ensures that no two edges incident on the same node
+    have the same color.
```
Removed the redundant sentence.

padath314 commented on networkx/algorithms/bipartite/edge_coloring.py:
```diff
@@ -0,0 +1,279 @@
+    strategy : str, optional
+        The strategy to use for edge coloring. Currently, the following strategies are supported :
+        "iterated-matching" :
+        "kempe-chain" :
```
Done.

padath314 commented on networkx/algorithms/bipartite/edge_coloring.py:
```diff
@@ -0,0 +1,279 @@
+            coloring[edge] = i
+
+        # Remove the edges in the matching from G1
+        G1.remove_edges_from(list(matching.items()))
```
This was oversight at our part. Rectified.

padath314 commented on networkx/algorithms/bipartite/edge_coloring.py:
```diff
@@ -0,0 +1,279 @@
+    for u, v in G.edges:
+        # Get the colors of edges ending at u and v
+        u_colors = set(used_colors[u].keys())
+        v_colors = set(used_colors[v].keys())
```
Done.

padath314 commented on networkx/algorithms/bipartite/edge_coloring.py:
```diff
@@ -0,0 +1,279 @@
+    B_major = G.subgraph(B_major_nodes)
+
+    top_nodes_A = top_nodes & A_major_nodes
+    top_nodes_B = top_nodes & B_major_nodes
```
This was an oversight at our part. The code has been optimised now.

padath314 commented:
In addition, we have also implemented edge coloring for bipartite multigraphs by making slight modifications to the existing code. We are seeking advice on whether to push these updates/modifications right now or wait until this pull request gets approved.

@dschult

dschult commented:
Yes!  Could you push those changes up now... That will likely reduce the wait for review of that later -- and not add much to the review time for this PR.   Thanks!

dschult commented on networkx/algorithms/bipartite/tests/test_edgecoloring.py:
```diff
@@ -0,0 +1,303 @@
+
+    # iterate through each edge in the graph
+
+    if isinstance(G, nx.MultiGraph):
```
```suggestion
    if G.is_multigraph():
```
I think  this idiom appears elsewhere too.

dschult commented on networkx/algorithms/bipartite/tests/test_edgecoloring.py:
```diff
@@ -0,0 +1,303 @@
+
+    def test_disconnected_exception(self, strategy):
+        edges = [(1, 2), (1, 3), (3, 4), (3, 5), (5, 6), (7, 9), (8, 9)]
+        G = nx.Graph()
+        G.add_edges_from(edges)
```
```suggestion
        G = nx.Graph(edges)
```
And a similar idiom elsewhere...

dschult commented on networkx/algorithms/bipartite/edge_colorings.py:
```diff
@@ -0,0 +1,326 @@
+""" Functions related to Edge Coloring"""
```
```suggestion
"""Functions related to Edge Coloring"""
```

dschult reviewed (commented):
This all looks good to me. I have just a couple nit-picky commets below and I'm ready to approve it.  Thanks for your persistence and patience!

Sreelaxme commented:
@dschult we have made all the suggested changes. circleci documentation test is failing now. Could you please help.

dschult commented:
The geospatial examples are failing -- definitely not related to this PR. And other PRs started having this trouble about the same time. Something changed in the CI's testing environment. We will need to track it down, but it is not a stopper for this PR.

dschult commented on networkx/algorithms/bipartite/edge_colorings.py:
```diff
@@ -0,0 +1,327 @@
+    # dictionary of dictionary
+    used_colors = {node: {} for node in G.nodes}
+
+    if G.is_multigraph():
+        edges = G.edges(keys=True)
+    else:
+        edges = G.edges
+
+    for edge in edges:
+        if G.is_multigraph():
+            u, v, key = edge
+        else:
+            u, v = edge
+            key = 0
```
```suggestion
    for edge in G.edges:
        u, v = edge[:2]
        key = edge[2] if G.is_multigraph() else 0
```

dschult commented on networkx/algorithms/bipartite/edge_colorings.py:
```diff
@@ -0,0 +1,327 @@
+        # Increment the color counter
+        i += 1
+
+    def convert_coloring(coloring1):
+        coloring2 = {}
+        # code to convert edge ->list to edge, key->list
+        for edge in coloring1:
+            u, v = edge
+            e_colors = coloring1[edge]
+            i = 0
+            for c in e_colors:
+                coloring2[(u, v, i)] = c
+                i = i + 1
+
+        return coloring2
```
```suggestion
    # convert edge ->list to edge, key->list
    converted_coloring = {}
    for (u, v), e_colors in coloring.items():
            for i, c in enumerate(e_colors):
                converted_coloring[(u, v, i)] = c

    return converted_coloring
```

dschult commented on networkx/algorithms/bipartite/edge_colorings.py:
```diff
@@ -0,0 +1,327 @@
+    coloring = {}
+
+    # Start coloring with i = 0 color
+    i = 0
```
it might be easier to read with this variable called `color` instead of `i`.  That's an opinion so I'll leave it up to you.

dschult commented on networkx/algorithms/bipartite/edge_colorings.py:
```diff
@@ -0,0 +1,327 @@
+    top_nodes_B = part_b & max_degree_nodes
+
+    # make two different graphs A_major and B_major
+    A_major_nodes = top_nodes_A | part_b
+    B_major_nodes = top_nodes_B | part_a
+
+    A_major = G.subgraph(A_major_nodes)
+    B_major = G.subgraph(B_major_nodes)
```
```suggestion
    A_major = G.subgraph(top_nodes_A | part_b)
    B_major = G.subgraph(top_nodes_B | part_a)
```

dschult commented on networkx/algorithms/bipartite/tests/test_edgecoloring.py:
```diff
@@ -0,0 +1,302 @@
+
+    # iterate through each edge in the graph
+
+    if G.is_multigraph():
+        edges = G.edges(keys=True)
+    else:
+        edges = G.edges()
+
+    for edge in edges:
+        if isinstance(G, nx.MultiGraph):
+            u, v, key = edge
+            e1 = (u, v, key)
+            e2 = (v, u, key)
+        else:
+            u, v = edge
+            e1 = (u, v)
+            e2 = (v, u)
```
```suggestion
    for e1 in G.edges:
        e2 = e1[1::-1] + (e1[2:] if G.is_multigraph() else ())
```

dschult reviewed (commented):
Here's another set of suggestions for simplifying the code.
Thanks very much. :)

dschult commented on networkx/algorithms/bipartite/tests/test_edgecoloring.py:
```diff
@@ -0,0 +1,302 @@
+            e2 = (v, u)
+
+        if e1 not in coloring:
+            return False
+        if e2 not in coloring:
+            return False
+        if coloring[e1] != coloring[e2]:
+            return False
+        else:
+            color = coloring[e1]
```
It would be nice to have the `if` clauses turned into `assert` statements.
That's so when they fail we can tell which clause is failing.

dschult reviewed (commented):
I had a partial review sitting for too long... Finally getting back to it.
Push back on anything that think should be done differently.

padath314 commented:
@dschult We are not able to figure out what is wrong. The same tests are failing for both this and PR #7397 .

dschult commented:
I think those errors have been fixed in the main branch. Can you rebase or merge the main branch into your branch and push that? That should update the code (which is independent of your changes) and allow the docs to be built. <fingers crossed>
